client/state/data-layer/wpcom/email-accounts/utils.js

```diff
@@ -0,0 +1,38 @@
+/**
+ * External Dependencies
+ */
+import { isArray, get, set } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import { convertToCamelCase } from 'state/data-layer/utils';
+
+/**
+ * Converts an array of email accounts into an object grouped by provider, and then domain.
+ *
+ * @param {Array} accounts list of accounts to convert
+ * @returns {object} a new object grouped as described above.
+ */
+const groupByProviderAndDomain = ( accounts ) => {
+	return accounts.reduce( ( aggregate, account ) => {
+		const path = [ account.providerSlug, account.domain ];
+		const state = get( aggregate, path, [] );
+		state.push( account );
+		return set( aggregate, path, state );
+	}, {} );
+};
```

Comment:
Do we really need to perform that grouping? What needs are we trying to address here? I suspect we will still display mailboxes ordered by domain name on the `Email` page but having this additional category level is going to make it more difficult to circle though mailboxes:

```javascript
emailAccounts: {
  '153108537': {
    accounts: {
      titan: {
        'domain.com': [
          {
            email: 'one@domain.com',
            siteId: 153108537,
            domain: 'domain.com',
            firstName: 'One',
            lastName: 'Last',
            productSlug: 'titan',
            productName: 'Titan Basic',
            providerSlug: 'titan'
          },
          {
            email: 'two@domain.com',
            siteId: 153108537,
            domain: 'domain.com',
            firstName: 'Two',
            lastName: 'Last',
            productSlug: 'titan',
            productName: 'Titan Basic',
            providerSlug: 'titan'
          }
        ]
      },
      gapps: {
        'other-domain.com': [
          {
            email: 'one@other-domain.com',
            siteId: 153108537,
            domain: 'other-domain.com',
            firstName: 'One',
            lastName: 'Last',
            productSlug: 'gapps',
            productName: 'G Suite Basic',
            providerSlug: 'gapps'
          }
        ]
      }
    },
    requesting: true,
    requestError: false
  }
}
```

We could simply rely on selectors to retrieve a list of Titan mailboxes if we really need to.

Reply:
I'd say let's go with whatever is the most likely format. Probably based on our expected DB schema (see my proposal in pbLl1t-3a-p2). So something like this:

```
account
   domain name
   product slug
   blog_id?
   mailboxes
       name
```

Using selectors we can easily filter out what we need - all accounts with product slug == Titan and blog_id = 1234, for example.

As I've mentioned, I want this PR to get rid of the boilerplate code - so let's settle on something that looks reasonable and then we can iterate in follow-up PRs.

Reply:
> Do we really need to perform that grouping? What needs are we trying to address here? I suspect we will still display mailboxes ordered by domain name on the `Email` page but having this additional category level is going to make it more difficult to circle though mailboxes

> We could simply rely on selectors to retrieve a list of Titan mailboxes if we really need to.

I have walked this back. It now closely matches our DB schema. I'll merge this and we can do further iterations later in subsequent PRs.